egs/wsj/s5/steps/dict/apply_g2p_phonetisaurus.sh

```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+# Copyright 2014  Johns Hopkins University (Author: Yenda Trmal)
+# Copyright 2016  Xiaohui Zhang
+#           2018  Ruizhe Huang
+# Apache 2.0
+
+# This script applies a trained Phonetisarus G2P model to
+# synthesize pronunciations for missing words (i.e., words in
+# transcripts but not the lexicon), and output the expanded lexicon.
+# The user could specify either nbest or pmass option 
+# to determine the number of output pronunciation variants, 
+# or use them together to get the intersection of two options.
+
+# Begin configuration section.  
+stage=0
+nbest=          # Generate up to $nbest variants
+pmass=          # Generate so many variants to produce $pmass ammount, like 90%, of the prob mass
+# End configuration section.
+
+echo "$0 $@"  # Print the command line for logging
+
+[ -f ./path.sh ] && . ./path.sh; # source the path.
+. utils/parse_options.sh || exit 1;
+
+set -u
+set -e
+
+if [ $# != 3 ]; then
+  echo "Usage: $0 [options] <g2p-model> <word-list> <lexicon-out>"
+  echo "... where <g2p-model> is the trained g2p model."
+  echo "          <word-list> is a list of words whose pronunciation is to be generated."
+  echo "          <lexicon-out> output lexicon, whose format is <word>\t<prob>\t<pronunciation> for each line."
+  echo "e.g.: $0 --nbest 1 exp/g2p/model.fst exp/g2p/oov_words.txt data/local/dict_nosp/lexicon.txt"
+  echo ""
+  echo "main options (for others, see top of script file)"
+  echo "  --nbest <int>    # Maximum number of hypotheses to produce. By default, nbest=20"
+  echo "  --pmass <float>  # Select the maximum number of hypotheses summing to a total mass of pmass amount, within [0, 1], for a word. By default, pmass=1.0"
+  exit 1;
+fi
+
+model=$1
+word_list=$2
+out_lexicon=$3
+
+[ ! -z $nbest ] && [[ ! $nbest =~ ^[1-9][0-9]*$ ]] && echo "$0: nbest should be a positive integer." && exit 1;
+[ -z $pmass ] && [ -z $nbest ] && echo "$0: nbest or/and pmass should be specified." && exit 1;
+
+if [ -z $pmass ]; then
```

Comment:
actually there doesn't seem to be a default.  I agree 20 is too much-- normally 3 would be a reasonable limit.

Reply:
Actually the reason we used 20 by default is that, when the user only sets pmass,  in order to get correct pron-probs, we need to specify an "nbest" value large enough, since phonetisaurus only computes pron-probs on the nbest list. When the user wants to rely on nbest, we always leave the responsibility to the user for setting the proper value, and that's why we didn't set it at the beginning. In summary, we want to allow all three ways of specifying constraints (pmass/nbest or both), and let the user to determine the proper values needed.

Reply:
Thanks! I will look into phonetisaurus codes again and make sure why we chose to set these default values. Will also consider how to make the code simpler.

Reply:
The code has been made simpler, but we keep the "nbest=20, pmass=1.0" stuff. The justification is as follows (an elaboration of what Xiaohui said above):

Users have three options here:

1) only set nbest, e.g. nbest=3
In this case, pmass needs to be implicitly set to 1.0 (instead of 0.0 by phonetisaurus's default), which would never affect our option nbest=3.

2) only set pmass, e.g. pmass=0.95
In this case, nbest is implicitly set to 20, because phonetisaurus computes probability over nbest list (PhonetisaurusScript.h: 166~186), thus we need to specify a large enough nbest value here.

3) set both nbest and pmass
In this case, the user takes the intersections of the two options and has full control of them.

What we meant by "default" was a bit misleading. They are actually some implicitly-set values, due to implementation reasons.